### PR TITLE
NMS-13378: Added Meta-Data search and custom information

### PR DESCRIPTION
**Pull request by christianpape**
* JIRA: https://issues.opennms.org/browse/NMS-13378

* Added search provider for meta-data values. 
* Introduced a new property `org.opennms.netmgt.search.info` which allows to add a custom line to search results. For instance you can define something like `org.opennms.netmgt.search.info=${node:sys-contact} (Department: ${asset:department})` to always display the sys-contact and the department below the node match.

### CODE REVIEW

**Comment by mfuhrmann** [comment]
@christianpape I've tried this branch without success. Here is what I've done:

Created a new properties file like this:
```
mf1045@mf1045-nb115:~/Documents/git/onms/stack-play/minimal-horizon$ cat container-fs/opt/opennms-overlay/etc/opennms.properties.d/searchbar.properties
org.opennms.netmgt.search.info=${node:stack} ${node:env} ${node:app)
```
The nodes I defined have Meta-Data like this
image

but ONMS throws an error while starting
```
	... 33 more
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'daoContext' defined in URL [jar:file:/opt/opennms/lib/opennms-dao-28.0.1-SNAPSHOT.jar!/beanRefContext.xml]: Bean instantiation via constructor failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.context.support.ClassPathXmlApplicationContext]: Constructor threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'sessionFactory' defined in class path resource [META-INF/opennms/applicationContext-shared.xml]: Invocation of init method failed; nested exception is java.lang.IllegalArgumentException: unmatched placeholder start [${node:stack} ${node:env} ${node:app)]
	at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:275)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1148)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1050)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:512)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
	at org.springframework.beans.factory.support.BeanDefinitionValueResolver.resolveReference(BeanDefinitionValueResolver.java:351)
	... 46 more
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.context.support.ClassPathXmlApplicationContext]: Constructor threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'sessionFactory' defined in class path resource [META-INF/opennms/applicationContext-shared.xml]: Invocation of init method failed; nested exception is java.lang.IllegalArgumentException: unmatched placeholder start [${node:stack} ${node:env} ${node:app)]
	at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:163)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:122)
	at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:267)
	... 55 more
```

When I remove 2 of the Meta-Data like this
```
mf1045@mf1045-nb115:~/Documents/git/onms/stack-play/minimal-horizon$ cat container-fs/opt/opennms-overlay/etc/opennms.properties.d/searchbar.properties
org.opennms.netmgt.search.info=${node:stack}
```

OpenNMS starts but the search box is broken: 
image

**Comment by christianpape** [comment]
Last bracket isn‘t a curly one in your expression 😎

File: core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/EntityScopeProviderImpl.java [hunk 1]
```diff
@@ -115,7 +115,6 @@ public Scope getScopeForNode(final Integer nodeId) {
             if (node.getAssetRecord() != null) {
                 Scope assetScope = new ObjectScope<>(Scope.ScopeName.NODE, node.getAssetRecord())
                         .map(ASSET, "category", (a) -> Optional.ofNullable(a.getCategory()))
-                        .map(ASSET, "vendor", (a) -> Optional.ofNullable(a.getVendor()))
```
Comment (fooker) [inline core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/EntityScopeProviderImpl.java:118]: Why was this removed?

**Review by fooker — CHANGES_REQUESTED** [review]

File: core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/EntityScopeProviderImpl.java (hunk 1 above)
Comment (christianpape) [reply core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/EntityScopeProviderImpl.java:118]: Because two lines later again the vendor is set.

**Review by fooker — APPROVED** [review]

**Comment by mfuhrmann** [comment]
@christianpape :facepalm: Sorry. Didn't see it.

Works perfectly now! Not sure where this red error message came from. Can't reproduce it.

And there is a cool side effect (or maybe designed behavior? ). The fact, that a non-existent variable won't be displayed brings up a different approach of how to combine variables.

In some cases, I have `node` and in some cases, I have `requisition` MetaData for the same matter (in this case `stack`, `env`, `app`). If you write it like this the tag always looks great :-)
`${node:stack}${requisition:stack}:${node:env}${requisition:env}:${node:app}${requisition:app}`

**Review by mfuhrmann — APPROVED** [review]
Should we have documentation for this new enhancement? I know there is no documentation about the search box at all. So maybe adding the docs is not an option now and within this PR.

But I can imagine, that this configuration enhancement never reaches out to the people if it is just a new property.

Let's ping @Bonrob2. Maybe she knows about plans to document the search panel in general? Or at least she now heard about this one and can refer to it somewhere.

**Comment by christianpape** [comment]
> @christianpape 🤦 Sorry. Didn't see it.
> 
> Works perfectly now! Not sure where this red error message came from. Can't reproduce it.
> 
> And there is a cool side effect (or maybe designed behavior? ). The fact, that a non-existent variable won't be displayed brings up a different approach of how to combine variables.
> 
> In some cases, I have `node` and in some cases, I have `requisition` MetaData for the same matter (in this case `stack`, `env`, `app`). If you write it like this the tag always looks great :-)
> `${node:stack}${requisition:stack}:${node:env}${requisition:env}:${node:app}${requisition:app}`

You can also use something like ${node:stack|requisition:stack}.

**Comment by mfuhrmann** [comment]
@christianpape Is this PR only about adding MetaData into the search panel? Or: is it also possible to remove for example the ForeignId and other default information?

**Comment by mfuhrmann** [comment]
> You can also use something like ${node:stack|requisition:stack}.

@christianpape That's cool. Is that a general notation that works everywhere? If not, we definitely should have docs about it. Or at least 1-2 examples in the properties file?

**Comment by fooker** [comment]
> > You can also use something like ${node:stack|requisition:stack}.
> 
> @christianpape That's cool. Is that a general notation that works everywhere? If not, we definitely should have docs about it. Or at least 1-2 examples in the properties file?

That works everywhere. See https://docs.opennms.com/horizon/28.0.0/operation/meta-data.html#ga-meta-data-dsl

**Comment by mfuhrmann** [comment]
> > > You can also use something like ${node:stack|requisition:stack}.
> > 
> > 
> > @christianpape That's cool. Is that a general notation that works everywhere? If not, we definitely should have docs about it. Or at least 1-2 examples in the properties file?
> 
> That works everywhere. See https://docs.opennms.com/horizon/28.0.0/operation/meta-data.html#ga-meta-data-dsl

@fooker Great!

**Comment by christianpape** [comment]
> @christianpape Is this PR only about adding MetaData into the search panel? Or: is it also possible to remove for example the ForeignId and other default information?

The central search displays matching nodes and the information which part of each node matched the given string. Each of the implemented search provider returns this information. For instance, an user can see wether an asset field, the node’s label or any of its IP addresses matched. This is a pretty nice functionality and I really don't see a reason to omit this information.

**Comment by mfuhrmann** [comment]
@christianpape I understand and you are right. It's a good behavior. In my setup the foreignId and nodeLabel are identical. So you get a result like:
image
This redundancy of information in these few pixels are confusing me :joy: Maybe that's why I thought it would be a cool option to hide it.
But of course, the foreignId and nodeLabel could be different and then it behaves different.
Thanks for clarifying.

**Comment by Bonrob2** [comment]
> Should we have documentation for this new enhancement? I know there is no documentation about the search box at all. So maybe adding the docs is not an option now and within this PR.
> 
> But I can imagine, that this configuration enhancement never reaches out to the people if it is just a new property.
> 
> Let's ping @Bonrob2. Maybe she knows about plans to document the search panel in general? Or at least she now heard about this one and can refer to it somewhere.

Yes, let's document the search panel in general and this in particular. I can create a Jira.
